### ISSUE TITLE
sql: Use hash-sharded indexes by default

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#78049]
Workloads involving sequential keys can have major performance problems on CockroachDB. Hash-sharded indexes alleviate this problem, but because they're not the default (and do not exist in other databases), many users who would benefit from them do not know to use them. Since hash-sharded indexes provide major performance benefits when they help, and have only minor drawbacks when they don't, we should consider using hash-sharded indexes by default, at least for data types that are prone to sequential access patterns.

Most conservatively, we could apply hash sharding by default only when the index starts with a SERIAL or TIMESTAMP column. Or we could be more aggressive about using hash sharding and use it with additional data types, when those data types are not the first column of the index, or even for all indexes (it's a little silly to hash-shard a UUID if we know it's being generated anew for each row, but that's not true of all UUID usage if it's in a multi-column index, so it could still be worthwhile).

Since hash-sharding an index does reduce scan performance (proportional to the number of buckets), we may want to use a smaller number of buckets by default than we would use for `USING HASH` without an explicit bucket count. For example, we may use 4 or 8 buckets by default for all columns, while an index that specifies `USING HASH` might continue to default to 16. (I'm not sure if this is worthwhile or if we should just use the same default unconditionally to simplify).

We'd also need a way to disable hash-sharded indexing when it is undesirable. We kind of have syntax for this now (specify a bucket count of 1), but that's not very obvious so if this is a common need we should consider a better way to specify it.

**Comment by awoods187** [comment]
This is a really cool idea!

**Comment by petermattis** [comment]
This is an interesting idea, though we have to be careful in that this could cause test/dev clusters to behave differently to production clusters. To make this concern concrete, we can't simply making hash-sharded indexes the default for new clusters as doing so could cause discrepancies between a new cluster used for testing an application, and a production cluster (which has been upgraded over time) used to run the application. This is a general problem with changing default behavior, not something specific to hash-sharded indexes.

**Comment by bdarnell** [comment]
Yeah, that's a valid concern. It's unfortunate that the de facto standard for schema management is to save the *source* form of your DDL statements and replay them, instead of either preprocessing them in some way to make defaults explicit (similar to the way our SHOW CREATE TABLES statement explicitly puts everything in one column family) or copying/branching from a source database. Another possible solution would be to add a parameter like rust's "edition" that would specify which version of the defaults you want to use, but again people would have to be trained to add that parameter to their migration scripts.

**Comment by glennfawcett** [comment]
Creating `HASH` indexes on `SERIAL` or PK's with a Sequence is certainly good as distributing the data, but I would caution about apply the same logic for `TIMESTAMP` data.  A HASH index on a `TIMESTAMP` column will distribute the workflow for `INSERT` but the same is not true for `SELECT` with range scans.... and the most common reason for an index on TIMESTAMP is to do a scan of a specific range of time.

**Comment by gnat** [comment]
Wanted to echo, very intriguing idea if we can address the migration story. This was a hangup for me and I already have hash sharded indexes as my default in my own database lib.

For `SERIAL` / `SEQUENCE` primary key this would be amazing to be on by default, as it would transparently handle most use cases.

**Comment by ajwerner** [comment]
> For SERIAL / SEQUENCE primary key this would be an amazing developer experience to be on by default, as it would transparently handle most use cases.

Out of curiousity, do you find that the ordering of `unique_rowid()` is important? One thought we've been having is to encourage broader use of `unordered_unique_rowid()` for integer key columns. It's like our `unique_rowid()` in that it is unique across the cluster without any coordination, but it rearranges the bits so that they are uniformly spread out as opposed to ordered.

**Comment by gnat** [comment]
If I understand the idea, unfortunately yes, because:

1. If I didn't need `unique_rowid()` I would just use UUID in the first place (or `INT8`/`unordered_unique_rowid()`).
2. It pushes the `SERIAL` / `SEQUENCE` story onto another key (or `TIMESTAMP`). This effectively just moves the issue to another column (which in turn will need to become a hash sharded index).

**Comment by ajwerner** [comment]
Cool, thanks. That makes a lot of sense. There are certainly cases where ordering is necessary and, in those cases, hash sharded indexes can do a good job scaling write load while preserving the ability to utilize that ordering. Sounds like you're in that scenario. The question I've been chewing on is: how common is that case? My current thinking is that a good bit of the time, folks are using an ordered default value only because it's the default. In those cases, a UUID might be better, but if the application is used to an integer, then it's harder to change it.

A possible outcome I'd be interested in is a new default default of `unordered_unique_rowid()` and then perhaps a session variable to choose an ordered default coupled with some degree of hash-sharding. I don't know if that's a good choice, but it seems better than the current defaults.

**Comment by gnat** [comment]
There's been many in-depth threads on use cases for sequential indexes for CRDB on the forum. Thankfully we at least have hash sharded indexes, but I'm always open to better developer experience.

>a new default of unordered_unique_rowid()

UUID everywhere is already the recommended pattern in CRDB for distributed indexes.

**Comment by gnat** [comment]
Side note, this would also automagically improve the [Cockroach Django backend](https://github.com/cockroachdb/django-cockroachdb), as there's currently no way to add `USING BUCKET` to an index :eyes:  Issue tracked here: https://github.com/cockroachdb/django-cockroachdb/issues/264

Just another soft advantage to add to the pile.

**Comment by gnat** [comment]
Now that I've tried using CRDB with a handfull of available ORM's / Migrators, I can say this would improve high performance compatibility with most (Tortoise, Piccolo, SQLalchemy) requiring less effort from Cockroach Labs and maintainers on those projects.
